Invalid retention policy rules can be created

**Describe the bug**

When creating a retention policy without specifying `tag_matching` (according to the provider documentation, this param is optional), the retention rule is successfully created but does not work.

```
2026-07-26T12:31:28Z [INFO] [/pkg/retention/job.go:101]: Load 5 candidates from repository common/gitlab-runner-helper
2026-07-26T12:31:28Z [ERROR] [/pkg/retention/job.go:246]: retention job: get selector by metadata: empty selector kind or decoration
```

Moreover - but it may irrelevant be to report this here - the Web interface is somehow broken (the first "broken" rule appears but the next ones do not, also the table loads forever).

<img width="1558" height="683" alt="Image" src="https://github.com/user-attachments/assets/dcdc8f40-b89b-4def-96d9-0d9a89cb027c" />

```
 ERROR TypeError: can't access property "trim", t is undefined
    getI18nKey https://harbor.lan.XX.fr/2842.121292f1fa2a8c0b.js:1
    getI18nKey https://harbor.lan.XX.fr/2842.121292f1fa2a8c0b.js:1
    ce https://harbor.lan.XX.fr/2842.121292f1fa2a8c0b.js:1
    Q4 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    R8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    s5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    i5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    o5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    R8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    s5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    i5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    o5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    R8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    s5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    i5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    ru https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    Z8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    R8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    s5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    i5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    o5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    R8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    s5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    i5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    ru https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    Z8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    R8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    s5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    i5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    o5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    R8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    s5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    i5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    ru https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    Z8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    R8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    s5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    i5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    o5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    R8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    s5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    i5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    ru https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    Z8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    R8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    s5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    i5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    o5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    R8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    s5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    i5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    ru https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    Z8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    R8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    s5 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    P8 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    Jo https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    synchronizeOnce https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    synchronize https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    tickImpl https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    _tick https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    invoke https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    onInvoke https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    invoke https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    run https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    run https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    _next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    g https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    emit https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    F https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    K2 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    onInvokeTask https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    invokeTask https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    runTask https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    invokeTask https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    Z https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    j https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    $ https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
main.1ee7e900e53b9d59.js:1:502471
    handleError https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    factory https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    invoke https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    run https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    runOutsideAngular https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    factory https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    invoke https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    onInvoke https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    invoke https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    run https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    run https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    _next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    g https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    next https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    emit https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    F https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    K2 https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    onInvokeTask https://harbor.lan.XX.fr/main.1ee7e900e53b9d59.js:1
    invokeTask https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    runTask https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    invokeTask https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    Z https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    j https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
    $ https://harbor.lan.XX.fr/polyfills.6524d45171bcf78c.js:1
```

**To Reproduce**

```
resource "harbor_retention_policy" "common" {
  scope    = harbor_project.common.id
  schedule = "Weekly"

  rule {
    most_recently_pushed = 2
    repo_matching        = "gitlab-runner-helper"
    untagged_artifacts   = false
  }

  ...
}
```

**Expected behavior**

I did not specify `tag_matching` because I thought, as it is optional, that it would match any tag.

One of `tag_matching` or `tag_excluding` should be required, or, when none of these settings is specified, `tag_matching` should be "*".

**Additional context**

- Provider Version v3.12.1
- Terraform v1.15.8
- Harbor Version v2.15.2-a97e7b83

--- a/docs/resources/retention_policy.md
+++ b/docs/resources/retention_policy.md
@@ -54,18 +54,27 @@ resource "harbor_retention_policy" "main" {
 
 ~> Multiple tags or repositories must be provided as a comma-separated list wrapped into curly brackets `{ }`. Otherwise, the value is interpreted as a single value.
 
-Optional:
+Required exactly one of retain parameters:
 
 - `always_retain` (Boolean) retain always.
-- `disabled` (Boolean) Specify if the rule is disable or not. Defaults to `false`
 - `most_recently_pulled` (Number) retain the most recently pulled n artifacts.
 - `most_recently_pushed` (Number) retain the most recently pushed n artifacts.
 - `n_days_since_last_pull` (Number) retains the artifacts pulled within the lasts n days.
 - `n_days_since_last_push` (Number) retains the artifacts pushed within the lasts n days.
+
+Required exactly one of repository parameters:
+
 - `repo_excluding` (String) For the repositories excluding.
 - `repo_matching` (String) For the repositories matching.
+
+Required exactly one of tag parameters:
+
 - `tag_excluding` (String) For the tag excluding.
 - `tag_matching` (String) For the tag matching.
+
+Optional:
+
+- `disabled` (Boolean) Specify if the rule is disable or not. Defaults to `false`
 - `untagged_artifacts` (Boolean) with untagged artifacts. Defaults to `true`
 
 ## Import

--- a/provider/resource_retention_policy.go
+++ b/provider/resource_retention_policy.go
@@ -1,13 +1,16 @@
 package provider
 
 import (
+	"context"
 	"encoding/json"
+	"fmt"
 	"log"
 	"strconv"
 	"strings"
 
 	"github.com/goharbor/terraform-provider-harbor/client"
 	"github.com/goharbor/terraform-provider-harbor/models"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/customdiff"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
 
@@ -36,50 +39,44 @@ func resourceRetention() *schema.Resource {
 							Optional: true,
 							Default:  false,
 						},
-						"n_days_since_last_pull": {
-							Type:     schema.TypeInt,
-							Optional: true,
-							Default:  nil,
-							// ConflictsWith: []string{"n_days_since_last_push"},
-						},
-						"n_days_since_last_push": {
-							Type:     schema.TypeInt,
-							Optional: true,
-							Default:  nil,
-							// ConflictsWith: []string{"n_days_since_last_pull"},
-						},
-						"most_recently_pulled": {
-							Type:     schema.TypeInt,
-							Optional: true,
-						},
-						"most_recently_pushed": {
-							Type:     schema.TypeInt,
-							Optional: true,
-						},
-						"always_retain": {
-							Type:     schema.TypeBool,
-							Optional: true,
-						},
-						"repo_matching": {
-							Type:     schema.TypeString,
-							Optional: true,
-							// ConflictsWith: []string{".repo_excluding"},
-						},
-						"repo_excluding": {
-							Type:     schema.TypeString,
-							Optional: true,
-							// ConflictsWith: []string{".repo_matching"},
-						},
-						"tag_matching": {
-							Type:     schema.TypeString,
-							Optional: true,
-							// ConflictsWith: []string{".tag_excluding"},
-						},
-						"tag_excluding": {
-							Type:     schema.TypeString,
-							Optional: true,
-							// ConflictsWith: []string{".tag_matching"},
-						},
+					"n_days_since_last_pull": {
+						Type:     schema.TypeInt,
+						Optional: true,
+						Default:  nil,
+					},
+					"n_days_since_last_push": {
+						Type:     schema.TypeInt,
+						Optional: true,
+						Default:  nil,
+					},
+					"most_recently_pulled": {
+						Type:     schema.TypeInt,
+						Optional: true,
+					},
+					"most_recently_pushed": {
+						Type:     schema.TypeInt,
+						Optional: true,
+					},
+					"always_retain": {
+						Type:     schema.TypeBool,
+						Optional: true,
+					},
+					"repo_matching": {
+						Type:     schema.TypeString,
+						Optional: true,
+					},
+					"repo_excluding": {
+						Type:     schema.TypeString,
+						Optional: true,
+					},
+					"tag_matching": {
+						Type:     schema.TypeString,
+						Optional: true,
+					},
+					"tag_excluding": {
+						Type:     schema.TypeString,
+						Optional: true,
+					},
 						"untagged_artifacts": {
 							Type:     schema.TypeBool,
 							Optional: true,
@@ -93,6 +90,9 @@ func resourceRetention() *schema.Resource {
 		Read:   resourceRetentionRead,
 		Update: resourceRetentionUpdate,
 		Delete: resourceRetentionDelete,
+		CustomizeDiff: customdiff.All(
+			retentionRuleParamValidation,
+		),
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
@@ -275,4 +275,69 @@ func resolveRules(model models.Retention) []interface{} {
 	}
 
 	return make([]interface{}, 0)
+}
+
+func retentionRuleParamValidation(_ context.Context, d *schema.ResourceDiff, _ interface{}) error {
+	rules := d.Get("rule").([]interface{})
+	paramFields := []string{"n_days_since_last_pull", "n_days_since_last_push", "most_recently_pulled", "most_recently_pushed", "always_retain"}
+	repoFields := []string{"repo_matching", "repo_excluding"}
+	tagFields := []string{"tag_matching", "tag_excluding"}
+
+	for i, raw := range rules {
+		rule := raw.(map[string]interface{})
+
+		// exactly one retain param
+		count := 0
+		for _, field := range paramFields {
+			v, ok := rule[field]
+			if !ok {
+				continue
+			}
+			switch val := v.(type) {
+			case bool:
+				if val {
+					count++
+				}
+			case int:
+				if val != 0 {
+					count++
+				}
+			}
+		}
+		if count == 0 {
+			return fmt.Errorf("rule[%d]: exactly one of %v must be set", i, paramFields)
+		}
+		if count > 1 {
+			return fmt.Errorf("rule[%d]: only one of %v can be set, got %d", i, paramFields, count)
+		}
+
+		// exactly one repo param
+		repoCount := 0
+		for _, field := range repoFields {
+			if v, ok := rule[field].(string); ok && v != "" {
+				repoCount++
+			}
+		}
+		if repoCount == 0 {
+			return fmt.Errorf("rule[%d]: exactly one of %v must be set", i, repoFields)
+		}
+		if repoCount > 1 {
+			return fmt.Errorf("rule[%d]: only one of %v can be set", i, repoFields)
+		}
+
+		// exactly one tag param
+		tagCount := 0
+		for _, field := range tagFields {
+			if v, ok := rule[field].(string); ok && v != "" {
+				tagCount++
+			}
+		}
+		if tagCount == 0 {
+			return fmt.Errorf("rule[%d]: exactly one of %v must be set", i, tagFields)
+		}
+		if tagCount > 1 {
+			return fmt.Errorf("rule[%d]: only one of %v can be set", i, tagFields)
+		}
+	}
+	return nil
 }

--- a/provider/resource_retention_policy.go
+++ b/provider/resource_retention_policy.go
@@ -39,44 +39,44 @@ func resourceRetention() *schema.Resource {
 							Optional: true,
 							Default:  false,
 						},
-					"n_days_since_last_pull": {
-						Type:     schema.TypeInt,
-						Optional: true,
-						Default:  nil,
-					},
-					"n_days_since_last_push": {
-						Type:     schema.TypeInt,
-						Optional: true,
-						Default:  nil,
-					},
-					"most_recently_pulled": {
-						Type:     schema.TypeInt,
-						Optional: true,
-					},
-					"most_recently_pushed": {
-						Type:     schema.TypeInt,
-						Optional: true,
-					},
-					"always_retain": {
-						Type:     schema.TypeBool,
-						Optional: true,
-					},
-					"repo_matching": {
-						Type:     schema.TypeString,
-						Optional: true,
-					},
-					"repo_excluding": {
-						Type:     schema.TypeString,
-						Optional: true,
-					},
-					"tag_matching": {
-						Type:     schema.TypeString,
-						Optional: true,
-					},
-					"tag_excluding": {
-						Type:     schema.TypeString,
-						Optional: true,
-					},
+						"n_days_since_last_pull": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  nil,
+						},
+						"n_days_since_last_push": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  nil,
+						},
+						"most_recently_pulled": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+						"most_recently_pushed": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+						"always_retain": {
+							Type:     schema.TypeBool,
+							Optional: true,
+						},
+						"repo_matching": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"repo_excluding": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"tag_matching": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"tag_excluding": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
 						"untagged_artifacts": {
 							Type:     schema.TypeBool,
 							Optional: true,
@@ -277,6 +277,47 @@ func resolveRules(model models.Retention) []interface{} {
 	return make([]interface{}, 0)
 }
 
+func countSetParamFields(rule map[string]interface{}, fields []string) int {
+	count := 0
+	for _, field := range fields {
+		v, ok := rule[field]
+		if !ok {
+			continue
+		}
+		switch val := v.(type) {
+		case bool:
+			if val {
+				count++
+			}
+		case int:
+			if val != 0 {
+				count++
+			}
+		}
+	}
+	return count
+}
+
+func countSetStringFields(rule map[string]interface{}, fields []string) int {
+	count := 0
+	for _, field := range fields {
+		if v, ok := rule[field].(string); ok && v != "" {
+			count++
+		}
+	}
+	return count
+}
+
+func validateExactlyOne(count, index int, fields []string) error {
+	if count == 0 {
+		return fmt.Errorf("rule[%d]: exactly one of %v must be set", index, fields)
+	}
+	if count > 1 {
+		return fmt.Errorf("rule[%d]: only one of %v can be set, got %d", index, fields, count)
+	}
+	return nil
+}
+
 func retentionRuleParamValidation(_ context.Context, d *schema.ResourceDiff, _ interface{}) error {
 	rules := d.Get("rule").([]interface{})
 	paramFields := []string{"n_days_since_last_pull", "n_days_since_last_push", "most_recently_pulled", "most_recently_pushed", "always_retain"}
@@ -285,58 +326,17 @@ func retentionRuleParamValidation(_ context.Context, d *schema.ResourceDiff, _ i
 
 	for i, raw := range rules {
 		rule := raw.(map[string]interface{})
-
-		// exactly one retain param
-		count := 0
-		for _, field := range paramFields {
-			v, ok := rule[field]
-			if !ok {
-				continue
-			}
-			switch val := v.(type) {
-			case bool:
-				if val {
-					count++
-				}
-			case int:
-				if val != 0 {
-					count++
-				}
-			}
+		// check that exactly one of the param fields is set
+		if err := validateExactlyOne(countSetParamFields(rule, paramFields), i, paramFields); err != nil {
+			return err
 		}
-		if count == 0 {
-			return fmt.Errorf("rule[%d]: exactly one of %v must be set", i, paramFields)
+		// check that exactly one of the repo fields is set
+		if err := validateExactlyOne(countSetStringFields(rule, repoFields), i, repoFields); err != nil {
+			return err
 		}
-		if count > 1 {
-			return fmt.Errorf("rule[%d]: only one of %v can be set, got %d", i, paramFields, count)
-		}
-
-		// exactly one repo param
-		repoCount := 0
-		for _, field := range repoFields {
-			if v, ok := rule[field].(string); ok && v != "" {
-				repoCount++
-			}
-		}
-		if repoCount == 0 {
-			return fmt.Errorf("rule[%d]: exactly one of %v must be set", i, repoFields)
-		}
-		if repoCount > 1 {
-			return fmt.Errorf("rule[%d]: only one of %v can be set", i, repoFields)
-		}
-
-		// exactly one tag param
-		tagCount := 0
-		for _, field := range tagFields {
-			if v, ok := rule[field].(string); ok && v != "" {
-				tagCount++
-			}
-		}
-		if tagCount == 0 {
-			return fmt.Errorf("rule[%d]: exactly one of %v must be set", i, tagFields)
-		}
-		if tagCount > 1 {
-			return fmt.Errorf("rule[%d]: only one of %v can be set", i, tagFields)
+		// check that exactly one of the tag fields is set
+		if err := validateExactlyOne(countSetStringFields(rule, tagFields), i, tagFields); err != nil {
+			return err
 		}
 	}
 	return nil

--- a/provider/resource_retention_policy_test.go
+++ b/provider/resource_retention_policy_test.go
@@ -2,6 +2,7 @@ package provider
 
 import (
 	"fmt"
+	"regexp"
 	"testing"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
@@ -126,6 +127,96 @@ func testAccCheckRetentionScheduleNone() string {
 		  }
 
 	  }
+	`)
+}
+
+func TestAccRetentionMissingTagSelector(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccCheckRetentionNoTagSelector(),
+				ExpectError: regexp.MustCompile(`one of`),
+			},
+		},
+	})
+}
+
+func testAccCheckRetentionNoTagSelector() string {
+	return fmt.Sprintf(`
+	resource "harbor_project" "main" {
+		name = "acctest_retention_no_tag"
+	}
+
+	resource "harbor_retention_policy" "main" {
+		scope    = harbor_project.main.id
+		schedule = "Daily"
+		rule {
+			n_days_since_last_pull = 5
+			repo_matching          = "**"
+		}
+	}
+	`)
+}
+
+func TestAccRetentionMissingRepoSelector(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccCheckRetentionNoRepoSelector(),
+				ExpectError: regexp.MustCompile(`one of`),
+			},
+		},
+	})
+}
+
+func testAccCheckRetentionNoRepoSelector() string {
+	return fmt.Sprintf(`
+	resource "harbor_project" "main" {
+		name = "acctest_retention_no_repo"
+	}
+
+	resource "harbor_retention_policy" "main" {
+		scope    = harbor_project.main.id
+		schedule = "Daily"
+		rule {
+			n_days_since_last_pull = 5
+			tag_matching           = "latest"
+		}
+	}
+	`)
+}
+
+func TestAccRetentionMissingRetainParam(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccCheckRetentionNoRetainParam(),
+				ExpectError: regexp.MustCompile(`one of`),
+			},
+		},
+	})
+}
+
+func testAccCheckRetentionNoRetainParam() string {
+	return fmt.Sprintf(`
+	resource "harbor_project" "main" {
+		name = "acctest_retention_no_param"
+	}
+
+	resource "harbor_retention_policy" "main" {
+		scope    = harbor_project.main.id
+		schedule = "Daily"
+		rule {
+			repo_matching = "**"
+			tag_matching  = "latest"
+		}
+	}
 	`)
 }
 

--- a/provider/resource_retention_policy_test.go
+++ b/provider/resource_retention_policy_test.go
@@ -220,6 +220,198 @@ func testAccCheckRetentionNoRetainParam() string {
 	`)
 }
 
+// ============================================================
+// Multi-rule valid tests
+// ============================================================
+
+func TestAccRetentionMultipleRulesValid(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckRetentionMultipleRulesValid(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckResourceExists(resourceHarborRetentionMain),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "schedule", "Daily"),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "rule.0.n_days_since_last_pull", "7"),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "rule.0.repo_matching", "**"),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "rule.0.tag_matching", "latest"),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "rule.0.disabled", "false"),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "rule.1.n_days_since_last_push", "14"),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "rule.1.repo_matching", "myrepo/**"),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "rule.1.tag_matching", "v*"),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "rule.1.disabled", "false"),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "rule.2.n_days_since_last_pull", "30"),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "rule.2.repo_matching", "archive/**"),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "rule.2.tag_matching", "*"),
+					resource.TestCheckResourceAttr(
+						resourceHarborRetentionMain, "rule.2.disabled", "true"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckRetentionMultipleRulesValid() string {
+	return fmt.Sprintf(`
+	resource "harbor_project" "main" {
+		name = "acctest_retention_multi"
+	}
+
+	resource "harbor_retention_policy" "main" {
+		scope    = harbor_project.main.id
+		schedule = "Daily"
+
+		rule {
+			n_days_since_last_pull = 7
+			repo_matching          = "**"
+			tag_matching           = "latest"
+		}
+
+		rule {
+			n_days_since_last_push = 14
+			repo_matching          = "myrepo/**"
+			tag_matching           = "v*"
+		}
+
+		rule {
+			n_days_since_last_pull = 30
+			repo_matching          = "archive/**"
+			tag_matching           = "*"
+			disabled               = true
+		}
+	}
+	`)
+}
+
+// ============================================================
+// Multi-rule invalid tests
+// ============================================================
+
+func TestAccRetentionMultiRuleMissingTagSelector(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccCheckRetentionMultiRuleMissingTag(),
+				ExpectError: regexp.MustCompile(`one of`),
+			},
+		},
+	})
+}
+
+func testAccCheckRetentionMultiRuleMissingTag() string {
+	return fmt.Sprintf(`
+	resource "harbor_project" "main" {
+		name = "acctest_retention_multi_notag"
+	}
+
+	resource "harbor_retention_policy" "main" {
+		scope    = harbor_project.main.id
+		schedule = "Daily"
+
+		rule {
+			n_days_since_last_pull = 7
+			repo_matching          = "**"
+			tag_matching           = "latest"
+		}
+
+		rule {
+			n_days_since_last_push = 10
+			repo_matching          = "**"
+		}
+	}
+	`)
+}
+
+func TestAccRetentionMultiRuleMissingRepoSelector(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccCheckRetentionMultiRuleMissingRepo(),
+				ExpectError: regexp.MustCompile(`one of`),
+			},
+		},
+	})
+}
+
+func testAccCheckRetentionMultiRuleMissingRepo() string {
+	return fmt.Sprintf(`
+	resource "harbor_project" "main" {
+		name = "acctest_retention_multi_norepo"
+	}
+
+	resource "harbor_retention_policy" "main" {
+		scope    = harbor_project.main.id
+		schedule = "Daily"
+
+		rule {
+			n_days_since_last_pull = 7
+			repo_matching          = "**"
+			tag_matching           = "latest"
+		}
+
+		rule {
+			n_days_since_last_push = 10
+			tag_matching           = "v*"
+		}
+	}
+	`)
+}
+
+func TestAccRetentionMultiRuleMissingRetainParam(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config:      testAccCheckRetentionMultiRuleMissingRetain(),
+				ExpectError: regexp.MustCompile(`one of`),
+			},
+		},
+	})
+}
+
+func testAccCheckRetentionMultiRuleMissingRetain() string {
+	return fmt.Sprintf(`
+	resource "harbor_project" "main" {
+		name = "acctest_retention_multi_noretain"
+	}
+
+	resource "harbor_retention_policy" "main" {
+		scope    = harbor_project.main.id
+		schedule = "Daily"
+
+		rule {
+			n_days_since_last_pull = 7
+			repo_matching          = "**"
+			tag_matching           = "latest"
+		}
+
+		rule {
+			repo_matching = "**"
+			tag_matching  = "v*"
+		}
+	}
+	`)
+}
+
 // func testReplicationPolicyDestinationNamespace() string {
 // 	return fmt.Sprintf(`
 // 	resource "harbor_project" "main" {

